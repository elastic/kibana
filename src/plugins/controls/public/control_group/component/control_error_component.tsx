/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { useState } from 'react';

import { EuiButtonEmpty, EuiPopover } from '@elastic/eui';
import { FormattedMessage, I18nProvider } from '@kbn/i18n-react';
import { Markdown } from '@kbn/kibana-react-plugin/public';

interface ControlErrorProps {
  error: Error | string;
}

export const ControlError = ({ error }: ControlErrorProps) => {
  const [isPopoverOpen, setPopoverOpen] = useState(false);
  const errorMessage = error instanceof Error ? error.message : error;

  const popoverButton = (
    <EuiButtonEmpty
      color="danger"
      iconSize="m"
      iconType="error"
      data-test-subj="control-frame-error"
      onClick={() => setPopoverOpen((open) => !open)}
      className={'errorEmbeddableCompact__button'}
      textProps={{ className: 'errorEmbeddableCompact__text' }}
    >
      <FormattedMessage
        id="controls.frame.error.message"
        defaultMessage="An error occurred. View more"
      />
    </EuiButtonEmpty>
  );

  return (
    <I18nProvider>
      <EuiPopover
        button={popoverButton}
        isOpen={isPopoverOpen}
        className="errorEmbeddableCompact__popover"
        anchorClassName="errorEmbeddableCompact__popoverAnchor"
        closePopover={() => setPopoverOpen(false)}
      >
        <Markdown
          markdown={errorMessage}
          openLinksInNewTab={true}
          data-test-subj="errorMessageMarkdown"
        />
      </EuiPopover>
    </I18nProvider>
  );
};
