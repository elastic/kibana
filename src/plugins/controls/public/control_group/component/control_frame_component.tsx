/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { useEffect, useMemo, useState } from 'react';
import classNames from 'classnames';
import {
  EuiButtonIcon,
  EuiFormControlLayout,
  EuiFormLabel,
  EuiFormRow,
  EuiIcon,
  EuiLink,
  EuiLoadingChart,
  EuiPopover,
  EuiText,
  EuiToolTip,
} from '@elastic/eui';

import { FormattedMessage } from '@kbn/i18n-react';
import { Markdown } from '@kbn/kibana-react-plugin/public';
import { useReduxContainerContext } from '@kbn/presentation-util-plugin/public';
import { ControlGroupReduxState } from '../types';
import { pluginServices } from '../../services';
import { EditControlButton } from '../editor/edit_control';
import { ControlGroupStrings } from '../control_group_strings';
import { useChildEmbeddable } from '../../hooks/use_child_embeddable';
import { TIME_SLIDER_CONTROL } from '../../../common';

interface ControlFrameErrorProps {
  error: Error;
}

const ControlFrameError = ({ error }: ControlFrameErrorProps) => {
  const [isPopoverOpen, setPopoverOpen] = useState(false);
  const popoverButton = (
    <EuiText className="errorEmbeddableCompact__button" size="xs">
      <EuiLink
        className="eui-textTruncate"
        color="subdued"
        onClick={() => setPopoverOpen((open) => !open)}
      >
        <EuiIcon type="alert" color="danger" />
        <FormattedMessage
          id="controls.frame.error.message"
          defaultMessage="An error has occurred. Read more"
        />
      </EuiLink>
    </EuiText>
  );

  return (
    <EuiPopover
      button={popoverButton}
      isOpen={isPopoverOpen}
      anchorClassName="errorEmbeddableCompact__popoverAnchor"
      closePopover={() => setPopoverOpen(false)}
    >
      <Markdown
        markdown={error.message}
        openLinksInNewTab={true}
        data-test-subj="errorMessageMarkdown"
      />
    </EuiPopover>
  );
};

export interface ControlFrameProps {
  customPrepend?: JSX.Element;
  enableActions?: boolean;
  embeddableId: string;
  embeddableType: string;
}

export const ControlFrame = ({
  customPrepend,
  enableActions,
  embeddableId,
  embeddableType,
}: ControlFrameProps) => {
  const embeddableRoot: React.RefObject<HTMLDivElement> = useMemo(() => React.createRef(), []);
  const [fatalError, setFatalError] = useState<Error>();

  const {
    useEmbeddableSelector: select,
    containerActions: { untilEmbeddableLoaded, removeEmbeddable },
  } = useReduxContainerContext<ControlGroupReduxState>();

  const controlStyle = select((state) => state.explicitInput.controlStyle);

  // Controls Services Context
  const {
    overlays: { openConfirm },
  } = pluginServices.getServices();

  const embeddable = useChildEmbeddable({ untilEmbeddableLoaded, embeddableId, embeddableType });

  const [title, setTitle] = useState<string>();

  const usingTwoLineLayout = controlStyle === 'twoLine';

  useEffect(() => {
    if (embeddableRoot.current) {
      embeddable?.render(embeddableRoot.current);
    }
    const inputSubscription = embeddable
      ?.getInput$()
      .subscribe((newInput) => setTitle(newInput.title));
    const errorSubscription = embeddable?.getOutput$().subscribe({
      error: setFatalError,
    });
    return () => {
      inputSubscription?.unsubscribe();
      errorSubscription?.unsubscribe();
    };
  }, [embeddable, embeddableRoot]);

  const floatingActions = (
    <div
      className={classNames('controlFrameFloatingActions', {
        'controlFrameFloatingActions--twoLine': usingTwoLineLayout,
        'controlFrameFloatingActions--oneLine': !usingTwoLineLayout,
      })}
    >
      {!fatalError && embeddableType !== TIME_SLIDER_CONTROL && (
        <EuiToolTip content={ControlGroupStrings.floatingActions.getEditButtonTitle()}>
          <EditControlButton embeddableId={embeddableId} />
        </EuiToolTip>
      )}
      <EuiToolTip content={ControlGroupStrings.floatingActions.getRemoveButtonTitle()}>
        <EuiButtonIcon
          data-test-subj={`control-action-${embeddableId}-delete`}
          aria-label={ControlGroupStrings.floatingActions.getRemoveButtonTitle()}
          onClick={() =>
            openConfirm(ControlGroupStrings.management.deleteControls.getSubtitle(), {
              confirmButtonText: ControlGroupStrings.management.deleteControls.getConfirm(),
              cancelButtonText: ControlGroupStrings.management.deleteControls.getCancel(),
              title: ControlGroupStrings.management.deleteControls.getDeleteTitle(),
              buttonColor: 'danger',
            }).then((confirmed) => {
              if (confirmed) {
                removeEmbeddable(embeddableId);
              }
            })
          }
          iconType="cross"
          color="danger"
        />
      </EuiToolTip>
    </div>
  );

  const embeddableParentClassNames = classNames('controlFrame__control', {
    'controlFrame--twoLine': controlStyle === 'twoLine',
    'controlFrame--oneLine': controlStyle === 'oneLine',
    'controlFrame--fatalError': !!fatalError,
  });

  function renderEmbeddablePrepend() {
    if (typeof embeddable?.renderPrepend === 'function') {
      return embeddable.renderPrepend();
    }

    return usingTwoLineLayout ? undefined : (
      <EuiToolTip anchorClassName="controlFrame__labelToolTip" content={title}>
        <EuiFormLabel className="controlFrame__formControlLayoutLabel" htmlFor={embeddableId}>
          {title}
        </EuiFormLabel>
      </EuiToolTip>
    );
  }

  const form = (
    <EuiFormControlLayout
      className={classNames('controlFrame__formControlLayout', {
        'controlFrameFormControlLayout--twoLine': controlStyle === 'twoLine',
      })}
      fullWidth
      prepend={
        <>
          {(embeddable && customPrepend) ?? null}
          {renderEmbeddablePrepend()}
        </>
      }
    >
      {embeddable && !fatalError && (
        <div
          className={embeddableParentClassNames}
          id={`controlFrame--${embeddableId}`}
          ref={embeddableRoot}
        >
          {fatalError && <ControlFrameError error={fatalError} />}
        </div>
      )}
      {fatalError && (
        <div className={embeddableParentClassNames} id={`controlFrame--${embeddableId}`}>
          {<ControlFrameError error={fatalError} />}
        </div>
      )}
      {!embeddable && (
        <div className={embeddableParentClassNames} id={`controlFrame--${embeddableId}`}>
          <div className="controlFrame--controlLoading">
            <EuiLoadingChart />
          </div>
        </div>
      )}
    </EuiFormControlLayout>
  );

  return (
    <>
      {embeddable && enableActions && floatingActions}
      <EuiFormRow
        data-test-subj="control-frame-title"
        fullWidth
        label={
          usingTwoLineLayout
            ? title || ControlGroupStrings.emptyState.getTwoLineLoadingTitle()
            : undefined
        }
      >
        {form}
      </EuiFormRow>
    </>
  );
};
