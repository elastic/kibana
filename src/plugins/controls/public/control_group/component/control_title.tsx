/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React from 'react';
import { EuiFormLabel, EuiPopover, EuiIcon, EuiToolTip, EuiLink } from '@elastic/eui';
import useToggle from 'react-use/lib/useToggle';
import { css } from '@emotion/react';
import { ControlInput } from '../../types';

const TitleWithPopoverMessage = ({
  title,
  helpMessage,
  embeddableId,
}: {
  title?: string;
  helpMessage: ControlInput['helpMessage'];
  embeddableId: string;
}) => {
  const [isPopoverOpen, togglePopover] = useToggle(false);

  return helpMessage?.text ? (
    <EuiFormLabel className="controlFrame__formControlLayoutLabel" htmlFor={embeddableId}>
      <>
        {title}
        <EuiPopover
          panelPaddingSize="s"
          button={
            <EuiIcon
              data-test-subj={`control-group-help-message-${embeddableId}`}
              type="iInCircle"
              size="m"
              onClick={togglePopover}
              css={css`
                cursor: pointer;
                margin: 0 2px 2px;
              `}
            />
          }
          isOpen={isPopoverOpen}
          offset={10}
          closePopover={() => togglePopover(false)}
          repositionOnScroll
          anchorPosition="upCenter"
          panelStyle={{ maxWidth: 250 }}
        >
          {helpMessage?.link ? (
            <>
              {helpMessage.text}{' '}
              <EuiLink
                data-test-subj={helpMessage.link['data-test-subj']}
                href={helpMessage.link.href}
                target="_blank"
              >
                {helpMessage.link.text}
              </EuiLink>
            </>
          ) : (
            <>{helpMessage.text}</>
          )}
        </EuiPopover>
      </>
    </EuiFormLabel>
  ) : null;
};

export const ControlTitle = ({
  title,
  helpMessage,
  embeddableId,
}: {
  helpMessage?: ControlInput['helpMessage'];
  title?: string;
  embeddableId: string;
}) => {
  return helpMessage ? (
    <TitleWithPopoverMessage title={title} helpMessage={helpMessage} embeddableId={embeddableId} />
  ) : (
    <EuiToolTip anchorClassName="controlFrame__labelToolTip" content={title}>
      <EuiFormLabel className="controlFrame__formControlLayoutLabel" htmlFor={embeddableId}>
        {title}
      </EuiFormLabel>
    </EuiToolTip>
  );
};
