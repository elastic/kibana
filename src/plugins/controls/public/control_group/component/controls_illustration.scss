@include euiBreakpoint('xs', 's') {
  .controlsIllustration {
    width: $euiSize * 6;
    height: $euiSize * 6;
  }
}
