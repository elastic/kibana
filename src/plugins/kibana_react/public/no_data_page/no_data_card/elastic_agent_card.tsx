/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/* eslint-disable @elastic/eui/href-or-on-click */

import React, { FunctionComponent } from 'react';
import { i18n } from '@kbn/i18n';
import { CoreStart } from 'kibana/public';
import { EuiButton, EuiCard, EuiCardProps } from '@elastic/eui';
import { useKibana } from '../../context';

export type ElasticAgentCardProps = Partial<EuiCardProps> & {
  solution: string;
  href: string;
  recommended?: boolean;
  buttonLabel?: string;
};

/**
 * Applies extra styling to a typical EuiAvatar
 */
export const ElasticAgentCard: FunctionComponent<ElasticAgentCardProps> = ({
  solution,
  recommended = true,
  href = 'app/integrations/browse',
  buttonLabel,
  ...cardRest
}) => {
  const {
    services: { http },
  } = useKibana<CoreStart>();
  const addBasePath = http.basePath.prepend;
  const basePathUrl = '/plugins/kibanaReact/assets/';

  return (
    // @ts-ignore
    <EuiCard
      paddingSize="l"
      href={href}
      title={i18n.translate('elasticAgentCard.title', {
        defaultMessage: 'Add integrations with Elastic Agent',
      })}
      description={i18n.translate('elasticAgentCard.title', {
        defaultMessage: `The Elastic Agent provides a simple, unified way to
        collect logs, metrics, and other types of
        data from your hosts.`,
      })}
      image={addBasePath(`${basePathUrl}elastic_agent_card.svg`)}
      betaBadgeLabel={recommended ? 'Recommended' : undefined}
      footer={
        // @ts-ignore
        <EuiButton
          href={href}
          onClick={cardRest?.onClick}
          target={cardRest?.target}
          fill={recommended}
          data-test-subj={`empty-page-agent-action`}
        >
          {buttonLabel ||
            i18n.translate('elasticAgentCard.buttonLabel', {
              defaultMessage: 'Find a {solution} integration',
              values: { solution },
            })}
        </EuiButton>
      }
      {...cardRest}
    />
  );
};
