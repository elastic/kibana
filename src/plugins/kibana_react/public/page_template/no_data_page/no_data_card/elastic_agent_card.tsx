/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import React, { FunctionComponent } from 'react';
import { i18n } from '@kbn/i18n';
import { CoreStart } from 'kibana/public';
import { EuiButton, EuiCard, EuiTextColor, EuiScreenReaderOnly } from '@elastic/eui';
import { useKibana } from '../../../context';
import { NoDataPageActions, NO_DATA_RECOMMENDED } from '../no_data_page';
import { RedirectAppLinks } from '../../../app_links';

export type ElasticAgentCardProps = NoDataPageActions & {
  solution: string;
};

/**
 * Applies extra styling to a typical EuiAvatar
 */
export const ElasticAgentCard: FunctionComponent<ElasticAgentCardProps> = ({
  solution,
  recommended,
  title,
  href,
  button,
  layout,
  category,
  ...cardRest
}) => {
  const {
    services: { http, application },
  } = useKibana<CoreStart>();
  const addBasePath = http.basePath.prepend;
  const image = addBasePath(`/plugins/kibanaReact/assets/elastic_agent_card.svg`);
  const canAccessFleet = application.capabilities.navLinks.integrations;
  const hasCategory = category ? `/${category}` : '';

  if (!canAccessFleet) {
    return (
      <EuiCard
        paddingSize="l"
        image={image}
        title={
          <EuiTextColor color="default">
            {i18n.translate('kibana-react.noDataPage.elasticAgentCard.noPermission.title', {
              defaultMessage: `Contact your administrator`,
            })}
          </EuiTextColor>
        }
        description={
          <EuiTextColor color="default">
            {i18n.translate('kibana-react.noDataPage.elasticAgentCard.noPermission.description', {
              defaultMessage: `This integration is not yet enabled. Your administrator has the required permissions to turn it on.`,
            })}
          </EuiTextColor>
        }
        isDisabled
      />
    );
  }

  const defaultCTAtitle = i18n.translate('kibana-react.noDataPage.elasticAgentCard.title', {
    defaultMessage: 'Add Elastic Agent',
  });

  const footer =
    typeof button !== 'string' && typeof button !== 'undefined' ? (
      button
    ) : (
      // The href and/or onClick are attached to the whole Card, so the button is just for show.
      // Do not add the behavior here too or else it will propogate through
      <EuiButton fill>{button || title || defaultCTAtitle}</EuiButton>
    );

  return (
    <RedirectAppLinks application={application}>
      <EuiCard
        paddingSize="l"
        image={image}
        href={href ?? addBasePath(`/app/integrations/browse${hasCategory}`)}
        // Bad hack to fix the need for an a11y title even though the button exists
        title={
          <EuiScreenReaderOnly>
            <span>{defaultCTAtitle}</span>
          </EuiScreenReaderOnly>
        }
        description={i18n.translate('kibana-react.noDataPage.elasticAgentCard.description', {
          defaultMessage: `Use Elastic Agent for a simple, unified way to collect data from your machines.`,
        })}
        betaBadgeProps={{ label: recommended ? NO_DATA_RECOMMENDED : undefined }}
        footer={footer}
        layout={layout as 'vertical' | undefined}
        {...cardRest}
      />
    </RedirectAppLinks>
  );
};
