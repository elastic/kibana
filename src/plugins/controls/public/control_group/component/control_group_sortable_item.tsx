/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { EuiFlexItem, EuiFormLabel, EuiIcon, EuiFlexGroup } from '@elastic/eui';
import React, { forwardRef, HTMLAttributes } from 'react';
import { useSortable } from '@dnd-kit/sortable';
import { CSS } from '@dnd-kit/utilities';
import classNames from 'classnames';

import { useReduxContainerContext } from '@kbn/presentation-util-plugin/public';
import { ControlFrame, ControlFrameProps } from './control_frame_component';
import { ControlGroupReduxState } from '../types';
import { ControlGroupStrings } from '../control_group_strings';

interface DragInfo {
  isOver?: boolean;
  isDragging?: boolean;
  draggingIndex?: number;
  index?: number;
}

export type SortableControlProps = ControlFrameProps & {
  dragInfo: DragInfo;
  isEditable: boolean;
};

/**
 * A sortable wrapper around the generic control frame.
 */
export const SortableControl = (frameProps: SortableControlProps) => {
  const { embeddableId, isEditable } = frameProps;
  const { over, listeners, isSorting, transform, transition, attributes, isDragging, setNodeRef } =
    useSortable({
      id: embeddableId,
      animateLayoutChanges: () => true,
      disabled: !isEditable,
    });

  frameProps.dragInfo = { ...frameProps.dragInfo, isOver: over?.id === embeddableId, isDragging };

  return (
    <SortableControlInner
      key={embeddableId}
      ref={setNodeRef}
      {...frameProps}
      {...attributes}
      {...listeners}
      style={{
        transition: transition ?? undefined,
        transform: isSorting ? undefined : CSS.Translate.toString(transform),
      }}
    />
  );
};

const SortableControlInner = forwardRef<
  HTMLButtonElement,
  SortableControlProps & { style: HTMLAttributes<HTMLButtonElement>['style'] }
>(
  (
    { embeddableId, embeddableType, dragInfo, style, isEditable, ...dragHandleProps },
    dragHandleRef
  ) => {
    const { isOver, isDragging, draggingIndex, index } = dragInfo;
    const { useEmbeddableSelector } = useReduxContainerContext<ControlGroupReduxState>();
    const panels = useEmbeddableSelector((state) => state.explicitInput.panels);

    const grow = panels[embeddableId].grow;
    const width = panels[embeddableId].width;
    const title = panels[embeddableId].explicitInput.title;

    const dragHandle = (
      <button
        ref={dragHandleRef}
        {...dragHandleProps}
        aria-label={`${ControlGroupStrings.ariaActions.getMoveControlButtonAction(title)}`}
        className="controlFrame__dragHandle"
      >
        <EuiIcon type="grabHorizontal" />
      </button>
    );

    return (
      <EuiFlexItem
        grow={grow}
        data-control-id={embeddableId}
        data-test-subj={`control-frame`}
        data-render-complete="true"
        className={classNames('controlFrameWrapper', {
          'controlFrameWrapper-isDragging': isDragging,
          'controlFrameWrapper-isEditable': isEditable,
          'controlFrameWrapper--small': width === 'small',
          'controlFrameWrapper--medium': width === 'medium',
          'controlFrameWrapper--large': width === 'large',
          'controlFrameWrapper--insertBefore': isOver && (index ?? -1) < (draggingIndex ?? -1),
          'controlFrameWrapper--insertAfter': isOver && (index ?? -1) > (draggingIndex ?? -1),
        })}
        style={style}
      >
        <ControlFrame
          enableActions={isEditable && draggingIndex === -1}
          embeddableId={embeddableId}
          embeddableType={embeddableType}
          customPrepend={isEditable ? dragHandle : undefined}
        />
      </EuiFlexItem>
    );
  }
);

/**
 * A simplified clone version of the control which is dragged. This version only shows
 * the title, because individual controls can be any size, and dragging a wide item
 * can be quite cumbersome.
 */
export const ControlClone = ({ draggingId }: { draggingId: string }) => {
  const { useEmbeddableSelector: select } = useReduxContainerContext<ControlGroupReduxState>();
  const panels = select((state) => state.explicitInput.panels);
  const controlStyle = select((state) => state.explicitInput.controlStyle);

  const width = panels[draggingId].width;
  const title = panels[draggingId].explicitInput.title;
  return (
    <EuiFlexItem
      className={classNames('controlFrameCloneWrapper', {
        'controlFrameCloneWrapper--small': width === 'small',
        'controlFrameCloneWrapper--medium': width === 'medium',
        'controlFrameCloneWrapper--large': width === 'large',
        'controlFrameCloneWrapper--twoLine': controlStyle === 'twoLine',
      })}
    >
      {controlStyle === 'twoLine' ? <EuiFormLabel>{title}</EuiFormLabel> : undefined}
      <EuiFlexGroup responsive={false} gutterSize="none" className={'controlFrame__draggable'}>
        <EuiFlexItem grow={false}>
          <EuiIcon type="grabHorizontal" className="controlFrame__dragHandle" />
        </EuiFlexItem>
        {controlStyle === 'oneLine' ? (
          <EuiFlexItem>
            <label className="controlFrameCloneWrapper__label">{title}</label>
          </EuiFlexItem>
        ) : undefined}
      </EuiFlexGroup>
    </EuiFlexItem>
  );
};
