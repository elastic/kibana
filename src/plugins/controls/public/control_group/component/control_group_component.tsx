/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import '../control_group.scss';

import { EuiButtonIcon, EuiFlexGroup, EuiFlexItem, EuiPanel } from '@elastic/eui';
import React, { useMemo, useState } from 'react';
import classNames from 'classnames';
import {
  arrayMove,
  SortableContext,
  rectSortingStrategy,
  sortableKeyboardCoordinates,
} from '@dnd-kit/sortable';
import {
  closestCenter,
  DndContext,
  DragEndEvent,
  DragOverlay,
  KeyboardSensor,
  PointerSensor,
  useSensor,
  useSensors,
  LayoutMeasuringStrategy,
} from '@dnd-kit/core';

import { ViewMode } from '@kbn/embeddable-plugin/public';
import { ControlClone, SortableControl } from './control_group_sortable_item';
import { useControlGroupContainerContext } from '../control_group_renderer';
import { ControlGroupStrings } from '../control_group_strings';

export const ControlGroup = () => {
  // Redux embeddable container Context
  const reduxContext = useControlGroupContainerContext();
  const {
    embeddableInstance: controlGroup,
    actions: { setControlOrders },
    useEmbeddableSelector: select,
    useEmbeddableDispatch,
  } = reduxContext;
  const dispatch = useEmbeddableDispatch();

  // current state
  const panels = select((state) => state.explicitInput.panels);
  const viewMode = select((state) => state.explicitInput.viewMode);
  const controlStyle = select((state) => state.explicitInput.controlStyle);
  const showAddButton = select((state) => state.componentState.showAddButton);

  const isEditable = viewMode === ViewMode.EDIT;

  const idsInOrder = useMemo(
    () =>
      Object.values(panels)
        .sort((a, b) => (a.order > b.order ? 1 : -1))
        .reduce((acc, panel) => {
          acc.push(panel.explicitInput.id);
          return acc;
        }, [] as string[]),
    [panels]
  );

  const [draggingId, setDraggingId] = useState<string | null>(null);
  const draggingIndex = useMemo(
    () => (draggingId ? idsInOrder.indexOf(draggingId) : -1),
    [idsInOrder, draggingId]
  );

  const sensors = useSensors(
    useSensor(PointerSensor),
    useSensor(KeyboardSensor, { coordinateGetter: sortableKeyboardCoordinates })
  );

  const onDragEnd = ({ over }: DragEndEvent) => {
    if (over) {
      const overIndex = idsInOrder.indexOf(over.id);
      if (draggingIndex !== overIndex) {
        const newIndex = overIndex;
        dispatch(setControlOrders({ ids: arrayMove([...idsInOrder], draggingIndex, newIndex) }));
      }
    }
    setDraggingId(null);
  };

  const emptyState = !(idsInOrder && idsInOrder.length > 0);
  // Empty, non-editable view is null
  if (!isEditable && emptyState) {
    return null;
  }

  let panelBg: 'transparent' | 'plain' | 'success' = 'transparent';
  if (emptyState) panelBg = 'plain';
  if (draggingId) panelBg = 'success';

  return (
    <>
      {idsInOrder.length > 0 || showAddButton ? (
        <EuiPanel
          borderRadius="m"
          color={panelBg}
          paddingSize={emptyState ? 's' : 'none'}
          data-test-subj="controls-group-wrapper"
          className={classNames('controlsWrapper', {
            'controlsWrapper--empty': emptyState,
            'controlsWrapper--twoLine': controlStyle === 'twoLine',
          })}
        >
          <EuiFlexGroup
            wrap={false}
            gutterSize="m"
            direction="row"
            responsive={false}
            alignItems="center"
            data-test-subj="controls-group"
          >
            <EuiFlexItem>
              <DndContext
                onDragStart={({ active }) => setDraggingId(active.id)}
                onDragEnd={onDragEnd}
                onDragCancel={() => setDraggingId(null)}
                sensors={sensors}
                collisionDetection={closestCenter}
                layoutMeasuring={{
                  strategy: LayoutMeasuringStrategy.Always,
                }}
              >
                <SortableContext items={idsInOrder} strategy={rectSortingStrategy}>
                  <EuiFlexGroup
                    className={classNames('controlGroup', {
                      'controlGroup-isDragging': draggingId,
                    })}
                    alignItems="center"
                    gutterSize="s"
                    wrap={true}
                  >
                    {idsInOrder.map(
                      (controlId, index) =>
                        panels[controlId] && (
                          <SortableControl
                            isEditable={isEditable}
                            dragInfo={{ index, draggingIndex }}
                            embeddableId={controlId}
                            embeddableType={panels[controlId].type}
                            key={controlId}
                          />
                        )
                    )}
                  </EuiFlexGroup>
                </SortableContext>
                <DragOverlay>
                  {draggingId ? <ControlClone draggingId={draggingId} /> : null}
                </DragOverlay>
              </DndContext>
            </EuiFlexItem>
            {showAddButton && (
              <EuiFlexItem grow={false}>
                <EuiButtonIcon
                  size="s"
                  iconSize="m"
                  display="base"
                  iconType={'plusInCircle'}
                  aria-label={ControlGroupStrings.management.getAddControlTitle()}
                  onClick={() => controlGroup.openAddDataControlFlyout()}
                />
              </EuiFlexItem>
            )}
          </EuiFlexGroup>
        </EuiPanel>
      ) : (
        <></>
      )}
    </>
  );
};
