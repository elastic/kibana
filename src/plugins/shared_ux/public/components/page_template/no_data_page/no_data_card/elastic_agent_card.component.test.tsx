/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

import { shallow } from 'enzyme';
import React from 'react';
import { ElasticAgentCardComponent } from './elastic_agent_card.component';

describe('ElasticAgentCard', () => {
  test('renders', () => {
    const component = shallow(<ElasticAgentCardComponent canAccessFleet={true} />);
    expect(component).toMatchSnapshot();
  });

  describe('props', () => {
    test('recommended', () => {
      const component = shallow(<ElasticAgentCardComponent recommended canAccessFleet={true} />);
      expect(component).toMatchSnapshot();
    });

    test('button', () => {
      const component = shallow(
        <ElasticAgentCardComponent button="Button" canAccessFleet={true} />
      );
      expect(component).toMatchSnapshot();
    });

    test('href', () => {
      const component = shallow(
        <ElasticAgentCardComponent button="Button" canAccessFleet={true} href={'some path'} />
      );
      expect(component).toMatchSnapshot();
    });

    test('category', () => {
      const component = shallow(
        <ElasticAgentCardComponent category="custom" canAccessFleet={true} />
      );
      expect(component).toMatchSnapshot();
    });
  });
});
