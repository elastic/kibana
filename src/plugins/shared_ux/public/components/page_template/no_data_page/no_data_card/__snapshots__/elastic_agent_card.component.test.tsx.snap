// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElasticAgentCard props button 1`] = `
<NoDataCard
  button="Button"
  description="Use Elastic Agent for a simple, unified way to collect data from your machines."
  image="test-file-stub"
  title="Add Elastic Agent"
/>
`;

exports[`ElasticAgentCard props category 1`] = `
<NoDataCard
  category="custom"
  description="Use Elastic Agent for a simple, unified way to collect data from your machines."
  image="test-file-stub"
  title="Add Elastic Agent"
/>
`;

exports[`ElasticAgentCard props href 1`] = `
<NoDataCard
  button="Button"
  description="Use Elastic Agent for a simple, unified way to collect data from your machines."
  href="some path"
  image="test-file-stub"
  title="Add Elastic Agent"
/>
`;

exports[`ElasticAgentCard props recommended 1`] = `
<NoDataCard
  description="Use Elastic Agent for a simple, unified way to collect data from your machines."
  image="test-file-stub"
  recommended={true}
  title="Add Elastic Agent"
/>
`;

exports[`ElasticAgentCard renders 1`] = `
<NoDataCard
  description="Use Elastic Agent for a simple, unified way to collect data from your machines."
  image="test-file-stub"
  title="Add Elastic Agent"
/>
`;
